### PR TITLE
Add capabilities for passing in cmdline arguments to clamscan and freshclam

### LINKED ISSUE #615: Malware scan fails to find EICAR string in image over 25Mb
**Version of Anchore Engine and Anchore CLI if applicable**:

```
anchore-cli --version
anchore-cli, version 0.8.1
kubectl describe pod anchore-demo-anchore-engine-analyzer-6c47f8766f-wrx6s | grep Image
    Image:         docker.io/anchore/anchore-engine:v0.8.1
    Image ID:      docker.io/anchore/anchore-engine@sha256:43e0a7fd25483c7b6d8889d892ac353d4e3f137a6c681b871269b67b2d4b5ec2
```

**What happened**:

I created a container to trial out various features of Anchore, it looks like:

```
FROM node:13

COPY credentials .
COPY package.json .
COPY index.js .
COPY node_modules/ node_modules
COPY virus .
```

Where `virus` is:

```
X5O!P%@AP[4\PZX54(P^)7CC)7}$EICAR-STANDARD-ANTIVIRUS-TEST-FILE!$H+H*
```

When I add the image, wait for it to complete and look at the malware content it gives:

```
anchore-cli image content 012345678910.dkr.ecr.ap-southeast-2.amazonaws.com/gareth-test:virus4 malware
Error: There is no image content (malware) to provide for 012345678910.dkr.ecr.ap-southeast-2.amazonaws.com/gareth-test:virus4
```

I turned up the logging to DEBUG to ensure it was actually scanning and it is indeed:

```
[service:worker] 2020-09-10 05:44:38+0000 [-] [Thread-168] [anchore_engine.clients.localanchore_standalone/run_anchore_analyzers()] [DEBUG] command succeeded: cmd=/usr/local/lib/python3.6/site-packages/anchore_engine/analyzers/modules/50_malware_scan.py /anchore_service f1960ece9712b2f3f6da633e051ba8f27ff11dc74ada9303d22f2563eb1880df /analysis_scratch/9bfee9d5-245c-4635-850e-58b40c57f6ea /analysis_scratch/9bfee9d5-245c-4635-850e-58b40c57f6ea/output /analysis_scratch/9bfee9d5-245c-4635-850e-58b40c57f6ea stdout=Beginning malware analyzer execution
[service:worker]        Malware scanner clamav enabled. Starting run
[service:worker]        running clamav scan
[service:worker]        Refreshing ClamAV DB
[service:worker]        Running command freshclam --stdout --datadir=/home/anchore/clamav/db --config-file=/home/anchore/clamav/freshclam.conf
[service:worker]        ClamAV scan metadata: {'db_version': {'daily': '25925', 'main': '59', 'bytecode': '331'}, 'db_update_enabled': True}
[service:worker]        Running command clamscan --suppress-ok-results --infected --recursive --allmatch --archive-verbose --tempdir=/analysis_scratch/9bfee9d5-245c-4635-850e-58b40c57f6ea --database=/home/anchore/clamav/db /analysis_scratch/9bfee9d5-245c-4635-850e-58b40c57f6ea/squashed.tar
[service:worker]        Raw ClamAV stdout: 
[service:worker]        ----------- SCAN SUMMARY -----------
[service:worker]        Known viruses: 8890276
[service:worker]        Engine version: 0.102.4
[service:worker]        Scanned directories: 0
[service:worker]        Scanned files: 1
[service:worker]        Infected files: 0
[service:worker]        Data scanned: 0.00 MB
[service:worker]        Data read: 967.80 MB (ratio 0.00:1)
[service:worker]        Time: 24.545 sec (0 m 24 s)
[service:worker]        ClamAV findings: []
[service:worker]        Analyzer output: {'malware': {'clamav': '{"scanner": "clamav", "findings": [], "metadata": {"db_version": {"daily": "25925", "main": "59", "bytecode": "331"}, "db_update_enabled": true}}'}}
[service:worker]        Handling output key malware stderr=
[service:worker] 2020-09-10 05:44:38+0000 [-] [Thread-168] [anchore_engine.clients.localanchore_standalone/run_anchore_analyzers()] [DEBUG] timing: specific analyzer time: /usr/local/lib/python3.6/site-packages/anchore_engine/analyzers/modules/50_malware_scan.py - 49.1749062538147
[service:worker] 2020-09-10 05:44:40+0000 [-] [Thread-168] [anchore_engine.clients.localanchore_standalone/analyze_image()] [DEBUG] timing: total analyzer time: 012345678910.dkr.ecr.ap-southeast-2.amazonaws.com/gareth-test@sha256:32b51c3024cb9e54bff58bfae413c005bb9b1d5ab92124ef98a02251b6abe70c - 196.2890224456787
```

During further troubleshooting I snagged the `squashed.tar` that it processes, loaded it into a container locally and ran the commands taken from the debug logs that correlated with what I saw in the code:

```
docker run -it -v $(pwd):/tmp/squased.tar docker.io/anchore/anchore-engine:v0.8.1 bash
freshclam --stdout --datadir=/home/anchore/clamav/db --config-file=/home/anchore/clamav/freshclam.conf
clamscan --suppress-ok-results --infected --recursive --allmatch --archive-verbose --tempdir=/analysis_scratch/9bfee9d5-245c-4635-850e-58b40c57f6ea --database=/home/anchore/clamav/db /tmp/squased.tar/squashed.tar 

----------- SCAN SUMMARY -----------
Known viruses: 8890276
Engine version: 0.102.4
Scanned directories: 0
Scanned files: 1
Infected files: 0
Data scanned: 0.00 MB
Data read: 967.80 MB (ratio 0.00:1)
Time: 16.878 sec (0 m 16 s)
```

It also did not find the EICAR string. *However*, when I run it on the untar'ed archive, it does find it:

```
clamscan --suppress-ok-results --infected --recursive --allmatch --archive-verbose  --database=/home/anchore/clamav/db /tmp/squased.tar/squashed/
/tmp/squased.tar/squashed/virus: Win.Test.EICAR_HDB-1 FOUND
/tmp/squased.tar/squashed/virus!(0): Win.Test.EICAR_HDB-1 FOUND

----------- SCAN SUMMARY -----------
Known viruses: 8890276
Engine version: 0.102.4
Scanned directories: 4045
Scanned files: 31645
Infected files: 1
Data scanned: 1362.94 MB
Data read: 854.34 MB (ratio 1.60:1)
Time: 204.778 sec (3 m 24 s)
```

When executing `clamscan` with `-v --debug` this shows up:

```
LibClamAV debug: cli_updatelimits: scansize exceeded (initial: 104857600, consumed: 0, needed: 1014814720)
LibClamAV debug: cli_updatelimits: filesize exceeded (allowed: 26214400, needed: 1014814720)
```

I believe the command needs to be expanded to include `--max-filesize` and `--max-scansize`, otherwise the maximum container size it will scan is 25Mb (or alternatively it could untar the file first), from the docs:

```
--max-filesize=#n
    Extract and scan at most #n kilobytes from each archive. You may pass the value in megabytes in format xM or xm, where x is a number. This option protects your system against DoS attacks (default: 25 MB, max: <4 GB) 
--max-scansize=#n
    Extract and scan at most #n kilobytes from each scanned file. You may pass the value in megabytes in format xM or xm, where x is a number. This option protects your system against DoS attacks (default: 100 MB, max: <4 GB) 
```

Unfortunately when I try to do this clamav throws errors:

```
clamscan --suppress-ok-results --infected --scan-archive=yes --recursive --allmatch --archive-verbose  --database=/home/anchore/clamav/db --max-filesize=4095m --max-scansize=4095M /tmp/squased.tar/squashed.tar
LibClamAV Warning: Bytecode run timed out in interpreter after 5000 opcodes
LibClamAV Warning: Bytecode 79 failed to run: CL_ETIMEOUT: Time limit reached
LibClamAV Error: cli_malloc(): Attempt to allocate 725456912 bytes. Please report to https://bugzilla.clamav.net
LibClamAV Warning: Couldn't grow the blob: we may be low on memory
LibClamAV Error: cli_malloc(): Attempt to allocate 725456912 bytes. Please report to https://bugzilla.clamav.net
LibClamAV Warning: Couldn't grow the blob: we may be low on memory
LibClamAV Error: cli_malloc(): Attempt to allocate 725456912 bytes. Please report to https://bugzilla.clamav.net
LibClamAV Warning: Couldn't grow the blob: we may be low on memory
LibClamAV Error: cli_malloc(): Attempt to allocate 725456912 bytes. Please report to https://bugzilla.clamav.net
LibClamAV Warning: Couldn't grow the blob: we may be low on memory
LibClamAV Error: cli_malloc(): Attempt to allocate 725456912 bytes. Please report to https://bugzilla.clamav.net
LibClamAV Warning: Couldn't grow the blob: we may be low on memory
LibClamAV Error: cli_malloc(): Attempt to allocate 725456912 bytes. Please report to https://bugzilla.clamav.net
LibClamAV Warning: Couldn't grow the blob: we may be low on memory
LibClamAV Error: cli_malloc(): Attempt to allocate 725456912 bytes. Please report to https://bugzilla.clamav.net
LibClamAV Warning: Couldn't grow the blob: we may be low on memory

----------- SCAN SUMMARY -----------
Known viruses: 8890276
Engine version: 0.102.4
Scanned directories: 0
Scanned files: 1
Infected files: 0
Data scanned: 1301.51 MB
Data read: 967.80 MB (ratio 1.34:1)
Time: 332.930 sec (5 m 32 s)
```

I have approximately 10Gb of free memory when I run this. My `squashed.tar` is just under 1Gb, which doesn't seem to be uncommon (given this is a trival node application in an official node base image).

I think there might be two options here:
1) untar it first
2) provide the max size options

I feel like the first is the better option but I know very little about Anchore and clamav, so I will leave this with you. :)

**What did you expect to happen**:

For EICAR string to be found

**Any relevant log output from /var/log/anchore**:

Above.

**What docker images are you using**:

Above

**How to reproduce the issue**:

Above.

### PR DESCRIPTION
**Pull request by dakaneye**
Signed-off-by: Samuel Dacanay <sam.dacanay@anchore.com>

**What this PR does / why we need it**: We need to make the clamscan command more configurable in case people need to process images larger than 25mb. 

**Which issue this PR fixes**: fixes #615

**Special notes**:

### CODE REVIEW

File: anchore_engine/analyzers/malware.py [hunk 1]
```diff
@@ -198,6 +206,7 @@ def __init__(self, config, squashed_tar_path, tempdir):
         self.enabled = self.config.get(self.name, {}).get('enabled', False) if self.config is not None and self.config.get(self.name) is not None and self.tar_path is not None else False
         self.refresh_enabled = self.config.get(self.name, {}).get('db_update_enabled', True) if self.config is not None and self.config.get(self.name) is not None else True
         self.tempdir = tempdir
+        self.cmdline_args = self.config.get(self.name, {}).get('cmdline_args', [])
```
Comment (dakaneye) [inline anchore_engine/analyzers/malware.py:209]: this could be renamed if anyone thinks of a better name! Maybe **clamscan_args**

**Comment by dakaneye** [comment]
Also noting that this needs to get updated to point to the release branch when it's created.

**Comment by dakaneye** [comment]
@zhill this is pointing to the 0.8.2-dev branch now, so should be ready for review.

**Review by zhill — CHANGES_REQUESTED** [review]
This allows the setting of parameters, but I don't see a new default for the size limits. I think a size limit of something like 10Gb (default docker image max size) is necessary to ensure the scans take place if configured and admins can tune it down.

If the image squashed.tar is larger than the setting value, the analyzer should log an error and skip the scan to ensure the results aren't incorrect.

File: anchore_engine/analyzers/malware.py [hunk 2]
```diff
@@ -33,17 +33,32 @@ def run_clamav(path_to_scan: str, tempdir: str, dbdir: str) -> str:
...
+    print('Running command {}'.format(clamscan_cmd))
+    status, output = subprocess.getstatusoutput(clamscan_cmd)
     if status in [CLAMAV_FINDING_EXITCODE, CLAMAV_NO_FINDING_EXITCODE]:
         return output
     else:
         raise ScanFailedException('clamav execution failed. returned exit code {}. Output = {}'.format(status, output))
 
 
-def refresh_clamavdb(databasedir: str, configfile: str) -> str:
+def append_cmdline_args(cmd: str, cmdline_args: list):
```
[the remark this answers is not in the document]
Comment (dakaneye) [reply anchore_engine/analyzers/malware.py]: added!

**Comment by zhill** [comment]
@dakaneye Was this based on master? I think it needs to be rebased onto v0.8.2-dev as it looks like its pulling in a bunch of other commits that aren't part of this fix.

File: anchore_engine/analyzers/malware.py [hunk 3]
```diff
@@ -5,7 +5,7 @@
 config_path = os.getenv('ANCHORE_FRESHCLAM_CONFIG_FILE', '/home/anchore/clamav/freshclam.conf')
 dbdir = os.getenv('ANCHORE_CLAMAV_DB_DIR', '/home/anchore/clamav/db')
 
-CLAMAV_CMD = "clamscan --suppress-ok-results --infected --recursive --allmatch --archive-verbose --tempdir={tempdir} --database={database}"
+CLAMAV_CMD = "clamscan --suppress-ok-results --infected --recursive --allmatch --archive-verbose --tempdir={tempdir} --database={database} --max-filesize=1000m --max-scansize=1000m"
```
Comment (zhill) [inline anchore_engine/analyzers/malware.py:8]: We discussed 10GB max, but this is 1GB.

**Review by zhill — CHANGES_REQUESTED** [review]
Default seems to differ from our discussion, but otherwise looks good.

File: anchore_engine/analyzers/malware.py (hunk 3 above)
Comment (dakaneye) [reply anchore_engine/analyzers/malware.py]: my mistake! Fixed to 4gb as discussed

**Pull request merged**